README.md
# Ama
Ama es una librería de utilidades hecha completamente en Python como parte de mi proyecto de Tesis de Grado para la Universidad Católica de Asunción. Ama realiza inferencias utilizando datos de radar GAMIC generados en el formato binario jerárquico HDF5, los cuales son convertidos a estructuras de datos nativas en Python, [pandas](https://pandas.pydata.org) y [numpy](http://www.numpy.org) utilizando la librería [wradlib](http://wradlib.org).

Ama es una aplicación ejecutable desde la línea de comandos. Para lanzarla ejecutar:

    $ python -m ama

## Ayuda
Para listar la ayuda de Ama, utilizar el siguiente comando::

    $ python -m ama --help

## Funciones
1. **Process Reflectivity:** Procesa un directorio de datos de forma recursiva y genera imágenes con mapas de reflectividad.
1. **Process Rainfall:** Procesa un directorio de datos de forma recursiva y genera imágenes con mapas de profundidad de lluvia.
1. **Correlate dBZ Location:** Procesa un archivo de datos y correlaciona dBZ con coordenadas geográficas.
1. **Show Data:** Muestra los datos en pantalla (DEBUG).
1. **Run:** Lanza un proceso que escucha por el ultimo archivo generado por el radar, lo procesa y lo insertado en una base de datos utilzando un servicio Web como interfaz.
1. **DBSCAN:** Agrupa reflectividades utilizando el algoritmo DBSCAN.

## Sitio Web
1. GitHub (<https://github.com/k-zen/Ama>)
1. Tesis de Grado / UCA (<http://apkc.net/_4>)

## Licencia
> Copyright (c) 2016, Andreas P. Koenzen <akc at apkc.net>
> All rights reserved.
>
> Redistribution and use in source and binary forms, with or without
> modification, are permitted provided that the following conditions are met:
>
> * Redistributions of source code must retain the above copyright notice, this
>   list of conditions and the following disclaimer.
> * Redistributions in binary form must reproduce the above copyright notice,
>   this list of conditions and the following disclaimer in the documentation
>   and/or other materials provided with the distribution.
>
> THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
> AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
> IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
> ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
> LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
> CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
> SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
> INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
> CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
> ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
> POSSIBILITY OF SUCH DAMAGE.


CHANGES.txt
=================== Release 0.1 01/05/2016 =====================
Descripción
    -- Primer versión de la librería.
Funciones:
    -- Procesar un directorio que contiene archivos con datos de radar y por cada archivo generar una
    imagen de las reflectividades (dbZ).
    -- Procesar un directorio que contiene archivos con datos de radar y por cada archivo generar una
    imagen de las intensidades de lluvia (mm/h).
    -- Procesar un directorio o un archivo único y correlacionar los decibeles de reflectividad a una
    ubicación geográfica determinada en coordenadas geográficas (Lat/Long). Puede generar un archivo
    resultante o llamar a un Web Service.
Cambios
    -- 19/07/2017
        * Agregado soporte para seguridad.
    -- 09/06/2017
        * Agregado soporte para detectar si el archivo a procesar es de *Polaridad Simple o Doble*.
    -- 30/05/2017
        * Correcciones menores.
    -- 29/05/2017
        * Agregado soporte para detectar clusters de tormenta a través del algoritmo DBSCAN.
    -- 20/05/2017
        * Cambios importantes para el funcionamiento. Removidos los filtros para las correlaciones y otros.
    -- 16/05/2017
        * Agregado soporte para procesar todos los archivos con extensión *mvol* y luego borrarlos.
        * Modificado tamaño mínimo de archivos que pueden ser procesados.
    -- 27/04/2017
        * Agregado soporte para escribir los datos de intensidad de lluvia en un archivo externo para luego
        usarlo en los filtros de ruido.
        * Agregado soporte para hasta 1 decimal en intensidad de lluvia.
    -- 25/04/2017
        * Modificado el minimo para el filtrado de intensidades de lluvia.
    -- 21/04/2017
        * Agregado soporte para timeout en llamada al Controlador y tambien soporte para excepciones.
    -- 06/04/2017
        * Agregado soporte para escuchar por archivos nuevos en un directorio. Estos archivos nuevos corresponden
        a datos de radar, los cuales son procesados en insertados en una BD Postgresql. Esta función se denomina
        *Run* o función *Run del Procesador*.
    -- 05/10/2016
        * Modificado la estructura del proyecto. Agregado soporte para *setuptools*.
        * Agregado soporte para funcion principal con parametros.
    -- 04/10/2016
        * Agregado nuevos metodos a la clase *Processor*.
    -- 29/09/2016
        * Agregado clase *Processor*, la cual seria el punto de entrada para el procesamiento de datos.

ama/utils.py
# -*- coding: utf-8 -*-

"""
Clase perteneciente al módulo de procesamiento de datos e inferencias Ama.

.. module:: utils
   :platform: Unix
   :synopsis: Funciones útiles varias.

.. moduleauthor:: Andreas P. Koenzen <akc@apkc.net>
"""

import os

__author__ = "Andreas P. Koenzen"
__copyright__ = "Copyright 2016, Proyecto de Tesis / Universidad Católica de Asunción."
__credits__ = "Andreas P. Koenzen"
__license__ = "BSD"
__version__ = "0.1"
__maintainer__ = "Andreas P. Koenzen"
__email__ = "akc@apkc.net"
__status__ = "Prototype"


class Colors:
    """
    Clase de colores para la terminal.
    """

    def __init__(self):
        pass

    BOLD = "\033[1m"
    WARNING = "\033[93m"
    FAIL = "\033[91m"
    ENDC = "\033[0m"


class Utils:
    """
    Clase utilitaria.
    """

    def __init__(self):
        pass

    @staticmethod
    def files_for_processing(origin, qt, file_size_limit):
        """
        Procesa un directorio y devuelve todos los archivos que deben ser
        procesados. Esta función devuelve los archivos de forma relativa
        desde la variable de entorno WRADLIB_DATA.

        :param origin: El directorio origen de datos.
        :param qt: La cantidad de archivos a procesar.
        :param file_size_limit: El tamaño máximo de un archivo a procesar. \
            Los archivos que sobrepasen el tamaño serán obviados.

        :return: Una lista con todos los archivos a procesar.
        """
        matches = []
        counter = 0

        for filename in os.listdir(origin):
            if filename.endswith(".mvol"):
                if qt != -1 and counter >= qt:
                    break

                full_path = os.path.join(origin, filename)
                relative_path = os.path.relpath(full_path, os.environ["WRADLIB_DATA"])

                if os.stat(full_path).st_size < file_size_limit:
                    if qt != -1:
                        counter += 1

                    matches.append(relative_path)

        return matches

    @staticmethod
    def deduplicate_correlated_data(ri, lat, lon, data):
        for i, (r, la, lo) in enumerate(data):
            if la == lat and lo == lon:
                print(Colors.WARNING + "WARN: Dato duplicado => RI={0:.2f}={1:.2f} Lat={2:.5f}={3:.5f} Long={4:.5f}={5:.5f}".format(
                    r, ri, la, lat, lo, lon) + Colors.ENDC)
                return False

        return True

    @staticmethod
    def should_process_file(filename, file_size_limit, double_polarization_mode):
        """
        Chequea si un archivo cumple con las condiciones para ser procesado.

        :param filename: El archivo a procesar.
        :param file_size_limit: El tamaño máximo de un archivo a procesar. \
            Los archivos que sobrepasen el tamaño serán obviados.
        :param double_polarization_mode: Si el modo del archivo es de "Polarización \
            Doble". En este modo los archivos son mayores en tamaño, debido a que \
            contienen varias elevaciones del Radar.

        :return: True si el archivo debe ser procesado, False de lo contrario.
        """

        import ama.processor as processor

        if filename.endswith(".mvol"):
            if double_polarization_mode == False:
                if os.stat(filename).st_size < file_size_limit:
                    return True
            else:
                #
                # Aplicar los chequeos.
                #
                if os.stat(filename).st_size > file_size_limit:
                    #
                    # Abrir el archivo y ver si es de hecho de "Polarización Doble".
                    #
                    print(Colors.BOLD + "INFO: Verificando si es de *Polarización Doble*..." + Colors.ENDC)
                    data, metadata = processor.Processor().process(filename, False)
                    counter = 0
                    for k in range(0, 11):
                        key = u"SCAN{0}".format(k)
                        if key in metadata:
                            counter += 1

                    #
                    # Solo si la cantidad de *Metadatos* es mayor que 1 el archivo puede ser de
                    # "Polarización Doble". De lo contrario significa que solo posee una capa de datos.
                    #
                    if counter > 1:
                        print(Colors.BOLD + "INFO: Archivo si es de *Polarización Doble*. Procesando..." + Colors.ENDC)
                        return True

        return False


ama/__main__.py
# -*- coding: utf-8 -*-

"""
==------------------------------------------==
==               Ama v0.1                   ==
==          http://apkc.net/_4              ==
== Autor: Andreas P. Koenzen <akc@apkc.net> ==
==------------------------------------------==

Uso:
====
ama [--process-reflectivity] [-t=target] [-d=destination]
    [--process-rainfall] [-t=target] [-d=destination]
    [--correlate-dbz-location] [-f=filename] [-d=destination] [-l=0] [--all] [--json-test]
    [--show-data] [-t=target]
    [--run] [-l=0]
    [--dbscan] [-f=filename] [-l=0] [--test]

Opciones:
=========
    --process-reflectivity   procesa un directorio de datos de forma recursiva
                             y genera imagenes de reflectividad
    --process-rainfall       procesa un directorio de datos de forma recursiva
                             y genera imagenes de profundidad de lluvia
    --correlate-dbz-location procesa un archivo de datos y correlaciona dBZ con
                             coordenadas geograficas
    --show-data              muestra los datos en pantalla (DEBUG)
    --run                    lanza un proceso que escucha por el ultimo archivo
                             generado por el radar, y lo procesa
    --dbscan                 procesa datos utilizando el algoritmo DBSCAN

    ---

    -t El PATH al directorio de datos de radar. Tomar en cuenta que se parte siempre
       desde la variable de entorno WRADLIB_DATA
    -d El PATH al directorio donde se guardaran los datos procesados. Tomar en cuenta
       que se parte siempre desde la variable de entorno AMA_EXPORT_DATA
    -f El nombre del archivo a procesar. Tomar en cuenta que se parte siempre
       desde la variable de entorno WRADLIB_DATA
    -l La capa de datos a utilizar. Valores correctos son: 0 hasta 10

    ---

    --all       Procesar todos los archivos en un directorio.
    --test      Habilitar modo pruebas/verificación.
    --json-test Habilitar modo pruebas/verificación con generación del archivo JSON.

Banderas:
=========
    --help muestra este mensaje de ayuda
"""

import ama.utils as utils
import ama.dbscan_processor as dbscan
import ama.processor as processor
import ama.file_listener as listener
import ama.show_data as show
import getopt
import os
import sys
import time

from watchdog.observers import Observer

__author__ = "Andreas P. Koenzen"
__copyright__ = "Copyright 2016, Proyecto de Tesis / Universidad Católica de Asunción."
__credits__ = "Andreas P. Koenzen"
__license__ = "BSD"
__version__ = "0.1"
__maintainer__ = "Andreas P. Koenzen"
__email__ = "akc@apkc.net"
__status__ = "Prototype"


class Usage(Exception):
    def __init__(self, msg):
        self.msg = msg


def usage():
    print(sys.exit(__doc__))


def main(argv=None):
    command = 0
    target = ""
    destination = ""
    filename = ""
    layer = 0
    process_all = False
    test = False
    json_test = False

    if argv is None:
        argv = sys.argv
    try:
        try:
            opts, args = getopt.getopt(
                argv[1:],
                "t:d:f:l:",
                [
                    "help",
                    "process-reflectivity",
                    "process-rainfall",
                    "correlate-dbz-location",
                    "show-data",
                    "run",
                    "all",
                    "test",
                    "json-test",
                    "dbscan"
                ]
            )
            if not opts:
                usage()
        except getopt.error, msg:
            raise Usage(msg)

        # DEBUG:
        # print opts

        for opt, arg in opts:
            if opt == "--help":
                usage()
            elif opt == "--process-reflectivity":
                command = 1
            elif opt == "--process-rainfall":
                command = 2
            elif opt == "--correlate-dbz-location":
                command = 3
            elif opt == "--show-data":
                command = 4
            elif opt == "--run":
                command = 5
            elif opt == "--dbscan":
                command = 6
            elif opt == "-t":
                target = arg
            elif opt == "-d":
                destination = arg
            elif opt == "-f":
                filename = arg
            elif opt == "-l":
                layer = int(arg)
            elif opt == "--all":
                process_all = True
            elif opt == "--test":
                test = True
            elif opt == "--json-test":
                json_test = True

        # tomar la decision.
        if command == 1:
            if not target and not destination:
                print(utils.Colors.FAIL + "ERROR: Origen y destino no definidos." + utils.Colors.ENDC)
                return 2

            processor.Processor().process_directory_generate_raw_images_from_reflectivity(target, destination)
        elif command == 2:
            if not target and not destination:
                print(utils.Colors.FAIL + "ERROR: Origen y destino no definidos." + utils.Colors.ENDC)
                return 2

            processor.Processor().process_directory_generate_raw_images_from_rainfall_intensity(target, destination)
        elif command == 3:
            if not filename and not destination:
                print(utils.Colors.FAIL + "ERROR: Nombre de archivo y destino no definidos." + utils.Colors.ENDC)
                return 2

            processor.Processor().correlate_dbz_to_location(filename, destination, process_all, layer, json_test)
        elif command == 4:
            if not target:
                print(utils.Colors.FAIL + "ERROR: Origen no definido." + utils.Colors.ENDC)
                return 2

            show.ShowData.show_data(target)
        elif command == 5:
            directory = os.path.join(os.environ["WRADLIB_DATA"], target)
            print(utils.Colors.BOLD + "INFO: Escuchando por adiciones en {0}.".format(directory) + utils.Colors.ENDC)

            event_handler = listener.FileListener(layer)
            observer = Observer()
            observer.schedule(event_handler, path=directory, recursive=False)
            observer.start()  # lanzar el proceso que observa adiciones en el directorio.
            try:
                while True:
                    time.sleep(1)
            except KeyboardInterrupt:
                observer.stop()  # agregar opcion de parar el observador con Ctrl+C.
            observer.join()
        elif command == 6:
            if not filename:
                print(utils.Colors.FAIL + "ERROR: Nombre de archivo no especificado." + utils.Colors.ENDC)
                return 2

            dbscan.DBSCANProcessor().plot_all_points(filename, layer, test)
    except Usage, err:
        print(utils.Colors.FAIL + "ERROR: {0}".format(err.msg) + utils.Colors.ENDC)
        print(utils.Colors.BOLD + "INFO: para ayuda utilizar --help" + utils.Colors.ENDC)
        return 2


if __name__ == "__main__":
    sys.exit(main())


ama/processor.py
# -*- coding: utf-8 -*-

"""
Clase perteneciente al módulo de procesamiento de datos e inferencias Ama.

.. module:: processor
   :platform: Unix
   :synopsis: Funciones útiles para el procesamiento de datos de radar.

.. moduleauthor:: Andreas P. Koenzen <akc@apkc.net>
"""

import ama.utils as utils
import base64
import matplotlib.pyplot as plt
import ntpath
import numpy as np
import pylab as pl
import os
import requests
import sys
import time
import wradlib as wrl

from haversine import haversine

__author__ = "Andreas P. Koenzen"
__copyright__ = "Copyright 2016, Proyecto de Tesis / Universidad Católica de Asunción."
__credits__ = "Andreas P. Koenzen"
__license__ = "BSD"
__version__ = "0.1"
__maintainer__ = "Andreas P. Koenzen"
__email__ = "akc@apkc.net"
__status__ = "Prototype"


class Processor:
    """
    Procesador de datos de radar. Clase principal donde se encuentran todas las funciones de
    procesamiento de datos.
    """

    def __init__(self):
        pass

    ###### OPCIONES MISCELANEAS ######
    DEBUG = False
    """
    boolean: Bandera para habilitar/deshabilitar modo DEBUG.
    """
    FILE_SIZE_LIMIT = 2 * 1024 * 1024  # 2MB para los archivos del GAMIC con zoom out.
    """
    int: Tamaño máximo de archivos a procesar. Todos los archivos que sobrepasen este tamaño serán obviados.
    """
    QT = -1
    """
    int: La cantidad de archivos a procesar dentro de un directorio.
    """
    SHOULD_REMOVE_PROCESSED_FILES = False
    """
    boolean: Bandera para habilitar el borrado de archivos luego de procesarlos. Solo para modo *run*.
    """

    ###### OPCIONES DE PROCESAMIENTO ######
    MINIMUM_REFLECTIVITY = 20.0
    """
    float: El valor mínimo para las reflectividades. Valores menores a este son obviados. 
    """
    MAXIMUM_REFLECTIVITY = 80.0
    """
    float: El valor máximo para las reflectividades. Valores mayores a este son obviados. 
    """

    @staticmethod
    def process(filename, show_info=True):
        """
        Procesa un archivo de datos de radar en formato GAMIC HDF5 y devuelve
        los datos.
        
        Esta función esta diseñada para archivos de "Polaridad Simple y Doble".

        :param filename: El nombre del archivo a procesar. El formato debe ser \
            *WRADLIB_DATA/<filename>*.
        :param show_info: Si debe mostrar mensajes de información o no.

        :return: Los datos de radar procesados.
        """
        filename = wrl.util.get_wradlib_data_file(filename)
        start = time.time()
        data, metadata = wrl.io.read_GAMIC_hdf5(filename)
        end = time.time()

        if show_info == 1:
            print(utils.Colors.BOLD + "INFO: Procesado \"{0}\" en {1} segundos.".format(filename, (end - start)) + utils.Colors.ENDC)

        return data, metadata

    def process_directory_generate_raw_images_from_reflectivity(self, origin, destination):
        """
        Procesa todos los archivos que se encuentran en el directorio
        de datos de forma recursiva y genera imagenes para cada set de datos
        utilizando los datos de reflectividad.
        
        Esta función esta diseñada para archivos de "Polaridad Simple".

        :param origin: El directorio origen de datos.
        :param destination: El directorio destino de las imagenes.

        :return: void
        """
        origin = os.path.join(os.environ["WRADLIB_DATA"], origin)
        destination = os.path.join(os.environ["AMA_EXPORT_DATA"], destination)
        matches = utils.Utils.files_for_processing(origin, self.QT, self.FILE_SIZE_LIMIT)

        if len(matches) > 0:
            for item in matches:
                data, metadata = Processor.process(item)
                fig = pl.figure(figsize=(10, 8))
                wrl.vis.plot_cg_ppi(data[u"SCAN0"][u"Z"]["data"], fig=fig)

                if not os.path.exists(destination):
                    print(utils.Colors.WARNING + "WARN: Destino no existe, creando ..." + utils.Colors.ENDC)
                    os.makedirs(destination)

                clean_filename = os.path.splitext(ntpath.basename(item))[0]
                pl.savefig(os.path.join(destination, (clean_filename + ".png")), bbox_inches="tight")
                plt.close(fig)

                if self.DEBUG == 1:
                    print(metadata)
                    print("------")
                    print(data)

                    for index, item in enumerate(matches):
                        print("{0} => {1}".format(index, item))

                    for index, item in enumerate(matches):
                        print("{0} => {1}".format(index, os.path.splitext(ntpath.basename(item))[0]))
        else:
            print(utils.Colors.FAIL + "ERROR: No hay archivos para procesar en *{0}*!".format(
                os.environ["WRADLIB_DATA"] + origin) + utils.Colors.ENDC)

    def process_directory_generate_raw_images_from_rainfall_intensity(self, origin, destination):
        """
        Procesa todos los archivos que se encuentran en el directorio
        de datos de forma recursiva y genera imagenes para cada set de datos
        utilizando los datos de intensidad de lluvia.

        Esta función esta diseñada para archivos de "Polaridad Simple".

        :param origin: El directorio origen de datos.
        :param destination: El directorio destino de las imagenes.

        :return: void
        """
        origin = os.path.join(os.environ["WRADLIB_DATA"], origin)
        destination = os.path.join(os.environ["AMA_EXPORT_DATA"], destination)
        matches = utils.Utils.files_for_processing(origin, self.QT, self.FILE_SIZE_LIMIT)

        if len(matches) > 0:
            for item in matches:
                data, metadata = Processor.process(item)

                Z = wrl.trafo.idecibel(data[u"SCAN0"][u"Z"]["data"])
                R = wrl.zr.z2r(Z, a=200., b=1.6)

                fig = pl.figure(figsize=(10, 8))
                ax, cf = wrl.vis.plot_ppi(R, cmap="spectral")
                pl.xlabel("Este del Radar (km)")
                pl.ylabel("Norte del Radar (km)")
                pl.title("Radar DINAC Fac. Veterinaria UNA\n6 min. profundidad de lluvia, " + metadata[u"SCAN0"]["Time"])
                cb = pl.colorbar(cf, shrink=0.8)
                cb.set_label("mm/h")
                pl.xlim(-128, 128)
                pl.ylim(-128, 128)
                pl.grid(color="grey")

                if not os.path.exists(destination):
                    print(utils.Colors.WARNING + "WARN: Destino no existe, creando ..." + utils.Colors.ENDC)
                    os.makedirs(destination)

                clean_filename = os.path.splitext(ntpath.basename(item))[0]
                pl.savefig(os.path.join(destination, (clean_filename + ".png")), bbox_inches="tight")
                plt.close(fig)

                if self.DEBUG == 1:
                    print(metadata)
                    print("------")
                    print(data)

                    for index, item in enumerate(matches):
                        print("{0} => {1}".format(index, item))

                    for index, item in enumerate(matches):
                        print("{0} => {1}".format(index, os.path.splitext(ntpath.basename(item))[0]))
        else:
            print(utils.Colors.FAIL + "ERROR: No hay archivos para procesar en *{0}*!".format(
                os.environ["WRADLIB_DATA"] + origin) + utils.Colors.ENDC)

    def single_correlate_dbz_to_location(self, filename, destination, layer):
        """
        Esta funcion realiza la correlacion entre dBZ y sus coordenadas geograficas en el mapa.
        
        Esta función esta diseñada para archivos de "Polaridad Simple y Doble".

        Formato del archivo a generar:
        ==============================
        Se genera un archivo *.ama, el cual no es nada mas que un archivo de texto separado por
        lineas, en el cual cada registro a su vez se encuentra separado por comas.

        Ejemplo:
            dBZ,latitude:longitude

        :param filename: El nombre del archivo a procesar.
        :param destination: El nombre del directorio en donde colocar los archivos resultantes.
        :param layer: La capa de datos a procesar. Cada capa corresponde a un ángulo de elevación del radar.

        :return: void
        """
        start = time.time()
        cdata = ""
        destination = os.path.join(os.environ["AMA_EXPORT_DATA"], destination,
                                   (os.path.splitext(ntpath.basename(filename))[0] + ".layer_{0}.ama".format(layer)))
        data, metadata = Processor.process(filename)
        clean_data = []
        layer_key = u"SCAN{0}".format(layer)

        file = open(destination, "w")

        radar_latitude = float(metadata["VOL"]["Latitude"])
        radar_longitude = float(metadata["VOL"]["Longitude"])

        for (row, column), value in np.ndenumerate(data[layer_key][u"Z"]["data"]):
            if value > -64.:
                rng = metadata[layer_key]["r"][column]
                azi = metadata[layer_key]["az"][row]
                dBZ = value
                lon, lat = wrl.georef.polar2lonlat(rng, azi, (radar_longitude, radar_latitude))

                # realizar los redondeos
                dBZ_value = float("{0:.1f}".format(dBZ))
                latitude_value = float("{0:.5f}".format(lat))
                longitude_value = float("{0:.5f}".format(lon))

                if dBZ_value >= self.MINIMUM_REFLECTIVITY and dBZ_value <= self.MAXIMUM_REFLECTIVITY:
                    clean_data.append((dBZ_value, latitude_value, longitude_value))

        for i, (dBZ, lat, lon) in enumerate(sorted(clean_data, key=lambda tup: tup[0])):
            line = "{0:.1f},{1:.5f}:{2:.5f}".format(dBZ, lat, lon)

            file.write(line + "\n")

            cdata += line

            if self.DEBUG == 1:
                print(line)

        file.close()

        end = time.time()

        print(utils.Colors.BOLD + "---" + utils.Colors.ENDC)
        print(utils.Colors.BOLD + "Tamaño Datos Enviados: {0}kb".format(sys.getsizeof(cdata) / 1024) + utils.Colors.ENDC)
        print(utils.Colors.BOLD + "Tiempo de Procesamiento: {0:.1f} minutos".format((end - start) / 60) + utils.Colors.ENDC)

    def correlate_dbz_to_location(self, filename, destination, process_all, layer, json_test=False):
        """
        Esta funcion procesa todo un directorio de archivos y por cada uno realiza la 
        correlacion entre dBZ y sus coordenadas geograficas en el mapa.
        
        Esta función esta diseñada para archivos de "Polaridad Simple y Doble".

        Formato del archivo a generar:
        ==============================
        Se genera un archivo *.ama, el cual no es nada mas que un archivo de texto separado por
        lineas, en el cual cada registro a su vez se encuentra separado por comas.

        Ejemplo:
            dBZ,latitude:longitude

        :param filename: El nombre del archivo a procesar.
        :param destination: El nombre del directorio en donde colocar los archivos resultantes.
        :param process_all Procesar todos los archivos en la carpeta donde se encuentra el archivo \
            pasado como *filename*.
        :param layer: La capa de datos a procesar. Cada capa corresponde a un ángulo de elevación del radar.
        :param json_test: Habilitar modo test/verificación para archivos JSON.

        :return: void
        """
        if process_all:
            origin = os.path.join(os.environ["WRADLIB_DATA"], os.path.split(filename)[0])
            destination = os.path.join(os.environ["AMA_EXPORT_DATA"], destination)
            matches = utils.Utils.files_for_processing(origin, self.QT, self.FILE_SIZE_LIMIT)

            if len(matches) > 0:
                for item in matches:
                    if utils.Utils.should_process_file(item, self.FILE_SIZE_LIMIT, True):
                        self.single_correlate_dbz_to_location(item, destination, layer)
        else:
            if json_test == 1:
                self.single_correlate_dbz_to_location_to_json(filename, layer, True)
            else:
                self.single_correlate_dbz_to_location(filename, destination, layer)

    def single_correlate_dbz_to_location_to_json(self, filename, layer, test=False):
        """
        Esta funcion realiza la correlacion entre dBZ y sus coordenadas geograficas en el mapa.
        
        Esta función esta diseñada para archivos de "Polaridad Simple y Doble".

        Formato del JSON a generar:
        ===========================
        {
            "fechaCarga": "2017-04-17T16:35:01.150Z",
            "notificar": True | False,
            "centroides": ["-25.23864:-57.52254", ..., "-25.23864:-57.52254"],
            "arrayDatos": ["6.0;-25.23864:-57.52254", ..., "7;-25.23864:-57.52254"]
        }

        :param filename: El nombre del archivo a procesar.
        :param layer: La capa de datos a procesar. Cada capa corresponde a un ángulo de elevación del radar.
        :param test: Habilitar modo test/verificación. En este modo no se llama al servicio Web solo \
            se genera el archivo de verificación.
        
        :return: void
        """

        import ama.dbscan_processor as dbscan

        ### DEBUG FILE
        debug_file = os.path.join(os.environ["AMA_EXPORT_DATA"], "JSON.debug")
        dfile = open(debug_file, "w")

        try:
            start = time.time()
            cdata = ""

            #
            # Datos que retorna el proceso de detección de clusters.
            #
            original, clustered, centroids, scan_time, radar_coordinates = dbscan.DBSCANProcessor().detect_dbz_clusters(filename, layer,
                                                                                                                        test)
            # construir el texto JSON.
            # fecha
            cdata += "{{\"fechaCarga\":\"{0}\",".format(scan_time)

            # detectar si cualquiera de los centroides está dentro del espacio radial de notificaciones.
            sendNotifications = False
            if len(centroids) > 0:
                for i, (centroid_lat, centroid_lon, centroid_dBZ) in enumerate(centroids):
                    if haversine((radar_coordinates[0], radar_coordinates[1]), (centroid_lat, centroid_lon)) <= 50:
                        sendNotifications = True
            cdata += "\"notificar\":\"{0}\",".format(sendNotifications)

            # agregar los centroides con sus coordenadas.
            cdata += "\"centroides\":["
            if len(centroids) > 0:
                for i, (centroid_lat, centroid_lon, centroid_dBZ) in enumerate(centroids):
                    line = "\"{0:.5f}:{1:.5f}\",".format(centroid_lat, centroid_lon)
                    # si es el último registro remover la coma al final.
                    if i == (len(centroids) - 1):
                        line = line[:-1]
                    # cuerpo
                    cdata += line
            cdata += "],"

            # agregar las dBZ con sus coordenadas.
            cdata += "\"arrayDatos\":["
            if len(clustered) > 0:
                for i, (lat, lon, dBZ) in enumerate(sorted(clustered, key=lambda tup: tup[2])):  # Ordenar por dBZ (índice 2).
                    line = "\"{0:.1f};{1:.5f}:{2:.5f}\",".format(dBZ, lat, lon)
                    # si es el último registro remover la coma al final.
                    if i == (len(clustered) - 1):
                        line = line[:-1]
                    # cuerpo
                    cdata += line
            cdata += "]}"

            if test == 1:
                dfile.write("{0}".format(cdata))

            # insertar los datos.
            # IMPORTANTE! Añadir timeout de 30 segundos para el pedido.
            if test == 0:
                url = "http://127.0.0.1:80/ama/datos/insertar"
                headers = {
                    'Content-type': 'application/json',
                    'Authorization': 'Basic ' + base64.b64encode('Ama@admin:amaPass2017')
                }
                response = requests.post(url, data=cdata, headers=headers, timeout=30)
                if response.status_code != requests.codes.ok:
                    print(utils.Colors.FAIL + "ERROR: Insertando datos en WS." + utils.Colors.ENDC)
                else:
                    print(utils.Colors.BOLD + "INFO: Datos insertados." + utils.Colors.ENDC)

            end = time.time()

            print(utils.Colors.BOLD + "---" + utils.Colors.ENDC)
            print(utils.Colors.BOLD + "Tamaño Datos Enviados: {0}kb".format(sys.getsizeof(cdata) / 1024) + utils.Colors.ENDC)
            print(utils.Colors.BOLD + "Tiempo de Procesamiento: {0:.1f} segundos".format((end - start)) + utils.Colors.ENDC)

            if len(clustered) == 0:
                print(utils.Colors.BOLD + "---" + utils.Colors.ENDC)
                print(utils.Colors.BOLD + "No se detectaron clusters. Se enviaron datos vacios al Controlador." + utils.Colors.ENDC)
                print(utils.Colors.BOLD + "Datos:" + utils.Colors.ENDC)
                print(utils.Colors.BOLD + "======" + utils.Colors.ENDC)
                print(utils.Colors.BOLD + "{0}".format(cdata) + utils.Colors.ENDC)
        except Exception as e:
            print(utils.Colors.FAIL + "ERROR: Corriendo trabajo de inserción." + utils.Colors.ENDC)
            print(utils.Colors.FAIL + "DESC: {0}".format(e) + utils.Colors.ENDC)
        finally:
            dfile.close()  # Siempre cerrar el archivo con datos de debug.

            # siempre borrar el archivo que fue procesado.
            if self.SHOULD_REMOVE_PROCESSED_FILES == 1:
                try:
                    os.remove(filename)
                except Exception as e:
                    print(utils.Colors.FAIL + "ERROR: Borrando archivo original." + utils.Colors.ENDC)
                    print(utils.Colors.FAIL + "DESC: {0}".format(e) + utils.Colors.ENDC)
